🚀 3단계 - 지하철 구간 관리

devMtn30 opened the pull request:
안녕하세요. 3단계 미션입니다.

devMtn30 committed:
feat: step1

devMtn30 committed:
feat: step2

devMtn30 committed:
feat: step2_노선관련api

devMtn30 committed:
fix: 충돌테스트

devMtn30 committed:
feat: 지하철 구간 관리 인수 조건 도출

devMtn30 committed:
feat: 지하철 구간 제거 인수 조건 도출

devMtn30 committed:
refactor: Station 테스트 코드 분리 및 함수분리

devMtn30 committed:
refactor: Line, 테스트 코드 분리 및 함수분리

devMtn30 committed:
refactor: Station 테스트 실패 원인 제거
- 요구사항을 잘못 읽어서 Line에 Station List 1:N 연관관계를 생성하였음, 이 부분 제거

devMtn30 committed:
refactor: 메소드 중복 제거,helper method public으로 변경

devMtn30 committed:
feat: 상행역이 종점역일 때 구간 등록 성공 인수 테스트

devMtn30 committed:
feat: section, 노선 인수 테스트

devMtn30 committed:
refactor: 메소드 분리

devMtn30 committed:
refactor: section 분리 메소드 helper로 이동

devMtn30 committed:
feat: Section Entity

devMtn30 committed:
feat: Section, Repository, Req, Res

devMtn30 committed:
refactor: section 추가로 인한 Station, Line수정

devMtn30 committed:
feat: section, 구간등록 성공테스트

devMtn30 committed:
feat: 구간 등록 인수 테스트 성공
 - 테스트 매개변수를 하드코딩으로 하다보니, ID가 변경되었을 때 실패함

devMtn30 committed:
feat: 구간 실패테스트 등록

devMtn30 committed:
fix: merge conflict line

devMtn30 committed:
Merge branch 'devmtn30' into step3

bactoria commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,154 @@
+     Then 에러 메시지가 반환된다.
+     */
+    @Test
+    @DisplayName("구간이 1개인 경우 역 제거 시 에러")
```
요구사항에 맞게 예외케이스 까지 꼼곰히 테스트 해 주셨네요 :100:

bactoria commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,68 @@
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    private Line line;
```
현재 Line > Sections > Section > Line > Sections ... 와 같이 순환관계가 만들어지는데요.

혹시 Section에 line이 추가된 이유가 있을까요?
꼭 필요한 상황이 아니라면 도메인 모델에서 의존성에 순환이 생기는 것은 좋은 모델링이 아니라고 생각해요. 😊

이런 상황에서는 우선 단방향으로 관계 설정을 한 후, 순환관계가 정말 필요한 상황이 오면 불가피하게 사용하면 어떨까요?
저는 아직은 실무에서 순환관계(양방향관계) 를 만들어 본 경험은 없어요.

devMtn30 replied:
@bactoria 리뷰 감사드립니다!  

생각해보지 못했던 부분이네요! 참고해서 수정해보겠습니다 :)  

추가로 궁금한게 있는데, 혹시 양방향 의존관계에서 문제점이 어떤게 있을까요?  

`JPA에서 양방향으로 연결된 Entity를 그대로 조회하는 경우 서로의 정보를 순환하면서 조회하다가 stackoverflow가 발생하게 된다.`  

의 경우에는 dto를 사용하기 때문에 애초에 발생하지 않는다고 생각하는데  
제가 생각하지 못한 다른 이유가 있을지 궁금합니다!

bactoria replied:
여기서는 Section에 있는 Line이 변경될 일은 없겠지만, 
변경된다고 한다면 아래 3가지 작업이 필요할 것 같아요.
1. 기존 Line에 있는 Section 제거
2. 변경할 Line에 Section 추가
3. Section에 있는 Line을 기존 Line에서 변경할 Line으로 변경 

여기서 3번은 양방향으로 인해 필요한 작업인데요.
Line과 Section이 양방향 관계인지 모르는 개발자는 1,2 과정의 작업을 할 거고, 이는 버그를 발생시킬 수 있어요.
즉, 양방향인지 아닌지 계속 신경쓰면서 개발을 해야한다는 의미인데 결국 개발 생산성을 떨어트릴 수 있어요. 🥲

여기서는 Section에 있는 Line이 변경되지 않으므로 그럼 양방향 설정을 해도 되겠지만
양방향 설정하기 전에는 Section에 Line 이 꼭 필요한가? 에 대한 필요성이 먼저 있어야 한다고 생각해요.

(추가적인 장점으로는 양방향을 제거하여 [ccdd14a](https://github.com/next-step/atdd-subway-map/pull/1124/commits/ccdd14aa64e68334e5d966a606b466ba6dd382b5) 읽어야 할 프로덕션 코드가 줄어들었네요. 😊 )

bactoria commented on src/main/java/subway/line/Line.java:
```diff
@@ -1,39 +1,35 @@
-    @Column(length = 20, nullable = false)
     private String name;
     private String color;
-    private Long upStationId;
-    private Long downStationId;
-    private Long distance;
```
이 3가지는 Section의 정보로 부터 조회할 수 있어 제거되었군요! 👍

bactoria commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,81 @@
+import javax.persistence.OneToMany;
+
+@Embeddable
+public class Sections {
```
일급컬렉션 👍 👍

bactoria commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private void validSection(Section section) {
+        if(sectionList.size() < 1) {
```
실제 어플리케이션에서는 노선을 생성할 때 구간이 무조건 1개가 존재할 텐데요!
현재 노선 생성 시 노선 객체를 구간없이 만들 수가 있는데, 구간이 없는 노선에 최초 구간을 추가하려고 이런 로직이 만들어진 것 네요.

처음부터 노선 객체 생성 시 구간이 1개 존재하도록 만들면 위와 같은 로직은 제거될 수 있을 것 같은데요.
실제로 노선 객체가 생성되기 위한 validation으로 구간이 1개 있어야 한다는 개념으로 접근 해 보면 도메인 모델이 더욱 견고해 질 것 같아요. 

꼭 반영은 하지 않아도 될 것 같은데, 이런 케이스에서 한번쯤 고민 해 보시면 좋을 것 같아 코멘트 달아두어요!

devMtn30 replied:
넵 감사합니다.  

실제로 노선이 생성될 때 구간이 필수로 1개가 필요하게 만드는 게 더 맞는 모델 설계인 것 같네요.  

감사합니다 :)

bactoria commented on src/main/java/subway/line/LineService.java:
```diff
@@ -49,6 +57,10 @@ public void deleteLine(Long id) {
     private LineResponse createLineResponse(Line line) {
-        return new LineResponse(line.getId(), line.getName(), line.getStations());
+        List<StationResponse> stationResponses = stationService.findAllStationById(
+            line.getFirstAndLastStationId());
```
이번 단계에서는 구간이 추가되었기 때문에 모든 역을 반환해야 하는데, 상행종점역과 하행종점역만 반환하는 버그가 생겼네요 😊
노선의 모든 역을 반환할 수 있도록 수정이 필요 해 보여요!

devMtn30 replied:
아.. 노선은 항상 상행종점역과 하행종점역만 가져오는걸로 알고 있었는데 아니었네요!  

수정 하겠습니다 :)

bactoria commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,68 @@
+    private Long downStationId;
+    private Long distance;
+
+    public boolean isSame(Section section) {
+        return downStationId.equals(section.getUpStationId()) || downStationId.equals(section.downStationId);
+    }
```
isSame이라는 메서드를 보았을 때는 equals와 유사한 메서드라고 생각을 하였는데요.
실제로는 파라미터로 받은 구간의 상행역과 하행역이 해당 구간의 하행역과 비교를 하고 있어서 어색하게 느껴졌어요.
그래서 첫번째로는 메서드 네이밍에 대한 수정이 필요 해 보이고, (적절한 메서드명을 짓기 어렵다면, 다르게 구현이 되는 것도 괜찮아 보여요!)

두번째로는 downStationId만 체크하고 있어, 노선의 첫번째 구간의 upStation에 대해서는 중복체크가 되지 않을 것 같은데, 
이는 2호선과 같은 순환노선에 대해 고려를 하신것인지 궁금해요! ☺️

devMtn30 replied:
이 개념이 헷갈렸네요 ㅜㅜ  

1번역, 2번역이 있을때 3번역을 추가하고 노선에 새로운 구간을 등록하게 된다면,  
현재 구간은 upstation: 1, downStation 2이고,  
추가되는 구간은 upStation: 2, downStation: 3  
이렇게 되니 새롭게 추가되는 upstation만 기존 노선에 비교하면 되도록 수정하였습니다!  

혹시 제가 잘못 이해한거라면 말씀해주세요!

bactoria replied:
- upStation: 상행역
- downStation: 하행역

**이미 해당 노선에 등록되어있는 역은 새로운 구간의 하행역이 될 수 없다.** 에 대한 구현이므로
새롭게 추가되는 **하행역**이 기존 노선에 존재하는지 비교해야 하는데요.

Section.isDuplicated(Section section) 에서는 새롭게 추가되는 **상행역**을 기존 노선에 비교하고 있어요.
이에 [1,2] 구간에 [2,2] 구간을 추가할 수 있는 버그가 있는데 하행역을 기존 노선에 비교하도록 수정이 필요 해 보여요! 😊

bactoria reviewed (changes requested):
이번 단계 꼼꼼히 잘 구현 해 주셨네요.
개발 프로세스도 잘 해주시고, 인수테스트도 가독성이 좋네요! 👍

지난 미션 때 커밋에 대한 리뷰를 드렸었는데, 
이번 단계에서는 의도에 맞게 커밋을 잘 분리 해 주셔서 
어떤 흐름으로 개발을 하셨는지 알게되어 덕분에 리뷰가 즐거웠어요. ☺️

몇가지 코멘트 남겨두었는데, 확인 후 리뷰요청 주세요!

devMtn30 committed:
refactor: 양방향 순환관계 수정

devMtn30 committed:
refactor: 노선이 생성될 때 구간이 필수로 1개가 필요하도록 변경

devMtn30 committed:
refactor: 노선 조회시 모든 역을 반환하도록 변경

devMtn30 committed:
refactor: section, 기존에 등록된 구간이 있는지 확인하는 메소드 조건 수정
1번역, 2번역이 있을때 3번역을 추가하고 노선에 새로운 구간을 등록하게 된다면,
현재 구간은 upstation: 1, downStation 2이고,
추가되는 구간은 upStation: 2, downStation: 3
이렇게 되니 새롭게 추가되는 upstation만 기존 노선에 비교

bactoria commented on src/main/java/subway/section/Sections.java:
```diff
@@ -0,0 +1,75 @@
+        return sectionList.size() > 1 && getLastDownStationId().equals(stationId);
+    }
+
+    public List<Long> getAllStationId() {
+        return sectionList.stream()
+                          .flatMap((section) -> {
+                              return Stream.of(section.getUpStationId(), section.getDownStationId());
+                          }).collect(Collectors.toList());
+    }
```
모든 역을 제공하도록 수정되었네요! 👍

현재 노선에 1,2,3 지하철역이 존재하는 경우 이 메서드의 응답은 [1, 2, 2, 3] 으로 내려오고 있는데,
[1, 2, 3] 이 내려오도록 중복이 제거되면 좋을 것 같아요~

bactoria reviewed (approved):
산님, 리뷰 반영 빠르게 잘 해주셨네요! 🙌
추가적으로 수정이 필요한 부분들에 대해 코멘트를 남겨두었으니 확인 부탁드립니다.

원활한 다음 미션 진행을 위해 이번 미션은 여기서 머지하도록 하겠습니다.
추가적으로 궁금하신 부분은 코멘트로 달아주시거나, 언제든 슬랙으로 말씀 주세요!
1주차 미션 하시느라 고생 많으셨어요. 👏
